crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -87,78 +170,128 @@ impl MemoryPool {
+            AutoEnabled::Auto => {
+                if mpk::is_supported() {
+                    mpk::keys()
+                } else {
+                    &[]
+                }
+            }
+            AutoEnabled::Enable => {
+                if mpk::is_supported() {
+                    mpk::keys()
+                } else {
+                    bail!("mpk is disabled on this system")
+                }
+            }
+            AutoEnabled::Disable => &[],
+        };
+
+        // This is a tricky bit of global state: when creating a memory pool
+        // that uses memory protection keys, we ensure here that any host code
+        // will have access to all keys (i.e., stripes). It's only when we enter
+        // the WebAssembly guest code (see `StoreInner::call_hook`) that we
+        // enforce which keys/stripes can be accessed. Be forewarned about the
+        // assumptions here:
+        // - we expect this "allow all" configuration to reset the default
+        //   process state (only allow key 0) _before_ any memories are accessed
+        // - and we expect no other code (e.g., host-side code) to modify this
+        //   global MPK configuration
+        if !pkeys.is_empty() {
+            mpk::allow(ProtectionMask::all());
+        }
+
         // Interpret the larger of the maximal size of memory or the static
         // memory bound as the size of the virtual address space reservation for
         // memory itself. Typically `static_memory_bound` is 4G which helps
         // elide most bounds checks in wasm. If `memory_pages` is larger,
-        // though, then this is a non-moving pooling allocator so create larger
-        // reservations for account for that.
-        let memory_size = config.limits.memory_pages.max(tunables.static_memory_bound)
+        // though, then this is a non-moving pooling allocator so we create
+        // larger reservations to account for that.
+        let max_memory_bytes = config.limits.memory_pages.max(tunables.static_memory_bound)
             * u64::from(WASM_PAGE_SIZE);
 
-        let memory_and_guard_size =
-            usize::try_from(memory_size + tunables.static_memory_offset_guard_size)
-                .map_err(|_| anyhow!("memory reservation size exceeds addressable memory"))?;
-
-        assert!(
-            memory_and_guard_size % crate::page_size() == 0,
-            "memory size {} is not a multiple of system page size",
-            memory_and_guard_size
-        );
-
-        let max_total_memories = config.limits.total_memories as usize;
-        let initial_memory_offset = if tunables.guard_before_linear_memory {
-            usize::try_from(tunables.static_memory_offset_guard_size).unwrap()
-        } else {
-            0
+        // Create a slab layout and allocate it as completely inaccessible
+        // region to start--`PROT_NONE`.
+        let constraints = SlabConstraints {
+            max_memory_bytes: max_memory_bytes as usize,
```

Comment:
Ok, I think I'm doing this more correctly in 52b245a7d. I'm internally using `expected_slot_bytes` as the name since it is more descriptive (to me) than `static_memory_bound`.